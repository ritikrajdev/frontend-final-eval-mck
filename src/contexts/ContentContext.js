import PropTypes from 'prop-types';
import React, { createContext } from 'react';

export const ContentContext = createContext([]);

export default function ContentContextProvider({ children }) {
  const [content, setContent] = React.useState(null);

  return (
    <ContentContext.Provider value={[content, setContent]}>
      {children}
    </ContentContext.Provider>
  );
}

ContentContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
