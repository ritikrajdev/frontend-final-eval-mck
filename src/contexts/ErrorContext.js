import PropTypes from 'prop-types';
import React, { createContext } from 'react';

export const ErrorContext = createContext([]);

export default function ErrorContextProvider({ children }) {
  const [error, setError] = React.useState(null);

  if (error) {
    setTimeout(() => {
      setError(null);
    }, 5000);
  }

  return (
    <ErrorContext.Provider value={[error, setError]}>
      {error && (
        <div
          style={{
            position: 'absolute',
            bottom: '20px',
            left: '20px',
            backgroundColor: 'white',
            display: 'inline-block',
            padding: '20px',
            borderRadius: '8px',
            color: 'red',
            border: '1px solid black',
          }}
        >
          {error}
        </div>
      )}
      {children}
    </ErrorContext.Provider>
  );
}

ErrorContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
