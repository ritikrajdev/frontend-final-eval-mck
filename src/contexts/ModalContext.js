import PropTypes from 'prop-types';
import React, { createContext } from 'react';
import Button from '../components/elements/Button';
import Input from '../components/elements/Input';

export const ModalContext = createContext([]);

export default function ModalContextProvider({ children }) {
  const [modal, setModal] = React.useState({
    // position: 'center' | 'right'
    show: false,
    position: 'center',
    title: 'Modal',
    fields: {},
    onSave: () => {
      return true;
    },
  });

  let positionStylingProps = {};

  if (modal.position === 'center') {
    positionStylingProps = {
      top: '50%',
      left: '50%',
      transform: 'translate(-50%, -50%)',
    };
  } else if (modal.position === 'right') {
    positionStylingProps = {
      top: '0',
      right: '0',
      bottom: '0',
    };
  }

  return (
    <ModalContext.Provider value={[modal, setModal]}>
      {modal.show && (
        <div
          style={{
            width: '100%',
            height: '100vh',
            backgroundColor: '#00000099',
            position: 'absolute',
          }}
          onClick={() => {
            setModal({ ...modal, show: false });
          }}
        >
          <form
            style={{
              ...positionStylingProps,
              position: 'absolute',
              backgroundColor: 'white',
              display: 'inline-block',
              padding: '20px',
              borderRadius: '8px',
              color: 'red',
              border: '1px solid black',
            }}
            onClick={(e) => e.stopPropagation()}
            onSubmit={(e) => {
              e.preventDefault();
              if (modal.onSave(modal.fields))
                setModal({ ...modal, show: false });
            }}
          >
            <h2>{modal.title}</h2>
            <br />
            {Object.entries(modal.fields).map(([key, value]) => (
              <div key={key}>
                <Input
                  label={key}
                  value={value}
                  onChange={(e) =>
                    setModal({
                      ...modal,
                      fields: { ...modal.fields, [key]: e.target.value },
                    })
                  }
                />
                <br />
              </div>
            ))}

            <center>
              <Button type='submit'>Save</Button>
            </center>
          </form>
        </div>
      )}
      {children}
    </ModalContext.Provider>
  );
}

ModalContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
