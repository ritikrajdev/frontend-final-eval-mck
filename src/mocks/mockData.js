export const mockCollections = [
  {
    id: 5,
    name: 'Laptops',
    schema: {
      brand: 'text',
    },
    userId: 1,
    formResponses: [
      {
        id: 8,
        form_id: 5,
        response: {
          brand: 'dell',
        },
      },
      {
        id: 4,
        form_id: 5,
        response: {
          brand: 'apple',
        },
      },
    ],
  },
];
