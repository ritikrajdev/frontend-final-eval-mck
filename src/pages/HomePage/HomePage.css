.home-page {
  display: flex;
  min-height: 100vh;
}

.cms-nav {
  background-color: var(--dark-gray);
  flex: 1;
}
ul {
  list-style: none;
}

.collections-containers {
  padding: 0 20px;
}

.cms-nav h1 {
  background-color: var(--dark-purple);
  color: white;
  padding: 20px 10px;
}

.cms-nav h3 {
  color: white;
}

.cms-nav li {
  color: #ffffff99;
}

.cms-content {
  background-color: var(--light-purple);
  flex: 3;
  display: flex;
  flex-direction: column;
}

.build-type-button {
  padding: 10px 20px;
}
