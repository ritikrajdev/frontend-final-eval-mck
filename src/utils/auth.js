export function getAuthToken() {
  return localStorage.getItem('authToken');
}

export function setAuthToken(token) {
  localStorage.setItem('authToken', token);
}

export function removeAuthToken() {
  localStorage.removeItem('authToken');
}
