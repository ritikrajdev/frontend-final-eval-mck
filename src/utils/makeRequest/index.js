import axios from 'axios';
import { ERROR_ROUTE, ERROR_WITH_CODE_ROUTE } from '../../constants/routes';
import { getAuthToken } from '../auth';

export async function makeRequest(apiEndpoint, dyamicConfig, navigate) {
  dyamicConfig = dyamicConfig ?? {};

  const authToken = getAuthToken();
  if (authToken) {
    if (dyamicConfig.headers) {
      dyamicConfig.headers['Authorization'] = authToken;
    } else {
      dyamicConfig.headers = { Authorization: authToken };
    }
  }

  try {
    const response = await axios({
      ...dyamicConfig,
      ...apiEndpoint,
    });

    return response.data;
  } catch (err) {
    if (navigate) {
      if (err.response?.status) {
        navigate(ERROR_WITH_CODE_ROUTE(err.response.status));
      } else {
        navigate(ERROR_ROUTE);
      }
    } else {
      throw err;
    }
  }
}
