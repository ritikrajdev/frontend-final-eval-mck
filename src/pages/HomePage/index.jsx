import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import Collections from '../../components/Collections';
import ContentTypes from '../../components/ContentTypes';
import { getAllFormsApiEndpoint } from '../../constants/apiEndpoints';
import { makeRequest } from '../../utils/makeRequest';

import './HomePage.css';

export default function HomePage() {
  const navigate = useNavigate();
  const [collections, setCollections] = useState(null);
  const [sideData, setSideData] = useState({
    // type: 'content' | 'collection',
    type: false,
    data: null,
  });

  useEffect(() => {
    makeRequest(getAllFormsApiEndpoint, {}, navigate).then((forms) => {
      setCollections(forms);
    });
  }, []);

  if (collections === null) {
    return 'Loading...';
  }

  return (
    <div className='home-page'>
      <div className='cms-nav'>
        <h1>CMS+</h1>
        <br />
        <div className='collections-containers'>
          <h3>COLLECTIONS</h3>
          <br />
          <ul>
            {collections.map((collection) => (
              <li key={collection.id}>â€¢ {collection.name}</li>
            ))}
          </ul>
          <br />
        </div>

        <h3
          className='build-type-button'
          style={{
            cursor: 'pointer',
          }}
          onClick={() => setSideData({ data: collections, type: 'content' })}
        >
          CONTENT TYPE BUILDER
        </h3>
      </div>
      <div className='cms-content'>
        {sideData.type &&
          (sideData.type === 'content' ? (
            <ContentTypes
              contentTypes={sideData.data}
              setContentTypes={(newData) => {
                setSideData({ ...sideData, data: newData });
              }}
            />
          ) : (
            <Collections />
          ))}
      </div>
    </div>
  );
}
