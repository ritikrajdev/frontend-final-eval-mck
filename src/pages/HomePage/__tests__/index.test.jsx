import { render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import HomePage from '..';
import { ContentContext } from '../../../contexts/ContentContext';
import { mockCollections } from '../../../mocks/mockData';
import { makeRequest } from '../../../utils/makeRequest';

jest.mock('../../../utils/makeRequest');

describe('HomePage', () => {
  it('should render the loading text', async () => {
    makeRequest.mockResolvedValue(mockCollections);
    const setCollections = jest.fn();
    render(
      <ContentContext.Provider value={[null, setCollections]}>
        <MemoryRouter>
          <HomePage />
        </MemoryRouter>
      </ContentContext.Provider>
    );
    expect(screen.getByText('Loading...')).toBeInTheDocument();
    await waitFor(() => {});
  });

  it('should render the collections', async () => {
    makeRequest.mockResolvedValue(mockCollections);
    const view = render(
      <ContentContext.Provider value={[mockCollections, jest.fn()]}>
        <MemoryRouter>
          <HomePage />
        </MemoryRouter>
      </ContentContext.Provider>
    );
    await waitFor(() => {});
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should render content type on content type builder click', async () => {
    makeRequest.mockResolvedValue(mockCollections);
    const view = render(
      <ContentContext.Provider value={[mockCollections, jest.fn()]}>
        <MemoryRouter>
          <HomePage />
        </MemoryRouter>
      </ContentContext.Provider>
    );
    await waitFor(() => {});
    const contentTypeBuilder = await screen.findByText('CONTENT TYPE BUILDER');
    contentTypeBuilder.click();
    expect(view.asFragment()).toMatchSnapshot();
    await waitFor(() => {});
  });

  it('should render collection on collections click', async () => {
    makeRequest.mockResolvedValue(mockCollections);
    const view = render(
      <ContentContext.Provider value={[mockCollections, jest.fn()]}>
        <MemoryRouter>
          <HomePage />
        </MemoryRouter>
      </ContentContext.Provider>
    );
    await waitFor(() => {});
    const contentTypeBuilder = await screen.findByText(
      'â€¢ ' + mockCollections[0].name
    );
    contentTypeBuilder.click();
    expect(view.asFragment()).toMatchSnapshot();
    await waitFor(() => {});
  });
});
