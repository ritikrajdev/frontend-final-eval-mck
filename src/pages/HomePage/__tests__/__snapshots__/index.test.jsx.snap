// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should render collection on collections click 1`] = `
<DocumentFragment>
  <div
    class="home-page"
  >
    <div
      class="cms-nav"
    >
      <h1>
        CMS+
      </h1>
      <br />
      <div
        class="collections-containers"
      >
        <h3>
          COLLECTIONS
        </h3>
        <br />
        <ul
          style="display: flex; flex-direction: column; gap: 8px; max-height: 45vh;"
        >
          <li
            style="cursor: pointer; background-color: transparent;"
          >
            • Laptops
          </li>
        </ul>
        <br />
      </div>
      <h3
        class="build-type-button"
        style="background-color: transparent; cursor: pointer;"
      >
        CONTENT TYPE BUILDER
      </h3>
    </div>
    <div
      class="cms-content"
    />
  </div>
</DocumentFragment>
`;

exports[`HomePage should render content type on content type builder click 1`] = `
<DocumentFragment>
  <div
    class="home-page"
  >
    <div
      class="cms-nav"
    >
      <h1>
        CMS+
      </h1>
      <br />
      <div
        class="collections-containers"
      >
        <h3>
          COLLECTIONS
        </h3>
        <br />
        <ul
          style="display: flex; flex-direction: column; gap: 8px; max-height: 45vh;"
        >
          <li
            style="cursor: pointer; background-color: transparent;"
          >
            • Laptops
          </li>
        </ul>
        <br />
      </div>
      <h3
        class="build-type-button"
        style="background-color: transparent; cursor: pointer;"
      >
        CONTENT TYPE BUILDER
      </h3>
    </div>
    <div
      class="cms-content"
    />
  </div>
</DocumentFragment>
`;

exports[`HomePage should render the collections 1`] = `
<DocumentFragment>
  <div
    class="home-page"
  >
    <div
      class="cms-nav"
    >
      <h1>
        CMS+
      </h1>
      <br />
      <div
        class="collections-containers"
      >
        <h3>
          COLLECTIONS
        </h3>
        <br />
        <ul
          style="display: flex; flex-direction: column; gap: 8px; max-height: 45vh;"
        >
          <li
            style="cursor: pointer; background-color: transparent;"
          >
            • Laptops
          </li>
        </ul>
        <br />
      </div>
      <h3
        class="build-type-button"
        style="background-color: transparent; cursor: pointer;"
      >
        CONTENT TYPE BUILDER
      </h3>
    </div>
    <div
      class="cms-content"
    />
  </div>
</DocumentFragment>
`;
